packages/services/hmi-server/src/main/java/software/uncharted/terarium/hmiserver/resources/DownloadResource.java

```diff
@@ -94,15 +96,82 @@ private byte[] getPDF(final String url) throws IOException, URISyntaxException {
 		return IOUtils.toByteArray(response.getEntity().getContent());
 	}
 
+	/**
+	 * Gets a PDF file from a given url
+	 *
+	 * @param url the url location (that may contain redirects)
+	 * @return the pdf file
+	 * @throws IOException
+	 * @throws URISyntaxException
+	 */
+	private String getPDFURl(final String url) throws IOException, URISyntaxException {
+		CloseableHttpClient httpclient = HttpClients.custom()
+				.disableRedirectHandling()
+				.build();
+
+		final HttpGet get = new HttpGet(url);
+		final HttpResponse response = httpclient.execute(get);
+
+		// Follow redirects until we actually get a document
+		if (response.getStatusLine().getStatusCode() >= 300 && response.getStatusLine().getStatusCode() <= 310) {
+			final String redirect = response.getFirstHeader("Location").getValue();
+			if (!redirect.startsWith("http")) {
+				return getPDFURl(normalizeRelativeUrl(redirect, url));
+			} else {
+				return getPDFURl(redirect);
+			}
+		} else {
+			// We actually have a document, if it's an HTML page with the content, look for
+			// a link to the pdf itself and follow
+			// it
+			final String contentType = response.getEntity().getContentType().getValue();
+			if (contentType.contains("html")) {
+				final String html = IOUtils.toString(response.getEntity().getContent(), StandardCharsets.UTF_8);
+				final Document document = Jsoup.parse(html);
+				final Elements links = document.select("a");
+				final String pdfUrl = links.stream()
+						.map(element -> element.attributes().get("href"))
+						.map(String::toLowerCase)
+						.filter(extractedUrl -> extractedUrl.endsWith(".pdf"))
+						.findFirst().orElse(null);
+
+				if (pdfUrl == null) {
+					return null;
+				}
+
+				if (!pdfUrl.startsWith("http")) {
+					final URI uri = new URI(url);
+					return getPDFURl(uri.getScheme() + "://" + uri.getHost() + pdfUrl);
+				} else {
+					return getPDFURl(pdfUrl);
+				}
+			}
+		}
+		return url;
+	}
+
 	@GET
 	@Produces(MediaType.APPLICATION_OCTET_STREAM)
 	public Response get(@QueryParam("doi") final String doi) throws IOException, URISyntaxException {
-
 		final byte[] pdfBytes = getPDF("https://unpaywall.org/" + doi);
```

Comment:
not required at all, but, if you ever want to do any sort of validation on a doi there is a regex expression used in ExtractionResource called `DOI_VALIDATION_PATTERN`